### PR TITLE
Add support for increasing an actor's MaxHP

### PR DESCRIPTION
**Pull request by reaperrr**
For RPG-like level systems, support for increasing max HP is pretty much a must-have.

I need this downstream and would like to avoid shipping a custom Mods.Common over this, plus other downstream projects might have a use for it as well.

### CODE REVIEW

File: OpenRA.Game/Traits/TraitsInterfaces.cs [hunk 1]
```diff
@@ -53,7 +53,7 @@ public sealed class DamageType { DamageType() { } }
 
 	public interface IHealthInfo : ITraitInfoInterface
 	{
-		int MaxHP { get; }
+		int DefaultMaxHP { get; }
```
Comment (Mailaender) [inline OpenRA.Game/Traits/TraitsInterfaces.cs]:
```suggestion
		int InitialMaxHitPoints { get; }
```
?

Comment (pchote) [reply OpenRA.Game/Traits/TraitsInterfaces.cs]: Do we really need to rename this at all?

Comment (reaperrr) [reply OpenRA.Game/Traits/TraitsInterfaces.cs]: 'Initial' might be better than 'Default', yeah.

I'm uncomfortable with HitPoints, though.
Not that I disagree with you (though HealthPoints might be more accurate), but if we went for consistency, we'd probably have to replace HP with HitPoints/HealthPoints everywhere in one go (including at yaml level, so with update rules and everything), and I feel that would be a bit excessive and a potential royal pain for a lot of modders and (mod) mappers for no practical benefit.
Basically, if it can be helped at all I'd like to avoid things snowballing out of scope over some minor plumbing change for a downstream feature.

Edit:

> Do we really need to rename this at all?

*shrugs* Maybe not. I mostly did that to reflect that this value is never updated even if actual MaxHP changes.

Comment (abcdefg30) [reply OpenRA.Game/Traits/TraitsInterfaces.cs]: Tbh, isn't this better solved by some sort of health modifier trait?

Comment (reaperrr) [reply OpenRA.Game/Traits/TraitsInterfaces.cs]: Assuming you mean an equivalent to `FirepowerMultiplier`, absolutely _no_, at least not for my usecase.

In my case, an actor can level up from 1 to 100, and its stats are supposed to increase by an exact flat amount each level, rather than a percentage.

Our current set of `*Multiplier` traits and their interfaces are completely unwieldy for that kind of purpose.

Though I guess using some kind of interface instead of modifiying MaxHP directly might still be more sensible.

Comment (pchote) [reply OpenRA.Game/Traits/TraitsInterfaces.cs]: IMO the approach you've taken here is the best solution, although i'm pretty sure that we will need to change `FrozenActor` to track the max as well as current HP to avoid problems with e.g. TD engineer capture vs sabotage. This would then also allow us to remove `IHealthInfo.MaxHP` and any potential confusion about its naming or use.

Comment (reaperrr) [reply OpenRA.Game/Traits/TraitsInterfaces.cs]: Done.

**Comment by pchote** [comment]
LGTM in principle, but not attempted to compile or test anything.

File: OpenRA.Game/Traits/TraitsInterfaces.cs [hunk 2]
```diff
@@ -51,10 +51,7 @@ public enum BlockedByActor
 	/// </summary>
 	public sealed class DamageType { DamageType() { } }
 
-	public interface IHealthInfo : ITraitInfoInterface
-	{
-		int MaxHP { get; }
-	}
+	public interface IHealthInfo : ITraitInfoInterface { }
```
Comment (abcdefg30) [inline OpenRA.Game/Traits/TraitsInterfaces.cs:54]: Is this still used now or are all usages gone?

File: OpenRA.Mods.Common/Traits/Health.cs [hunk 3]
```diff
@@ -125,6 +123,15 @@ void INotifyOwnerChanged.OnOwnerChanged(Actor self, Player oldOwner, Player newO
 			notifyKilledPlayer = newOwner.PlayerActor.TraitsImplementing<INotifyKilled>().ToArray();
 		}
 
+		public void ModifyHP(int addHP, int addMaxHP)
+		{
+			if (IsDead)
+				return;
+
+			hp += addHP;
+			MaxHP += addMaxHP;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Health.cs:132]: We probably should make sure that `hp` isn't increased above the (new) `MaxHP` value.

**Comment by PunkPun** [comment]
Adding to Future milestone. The engine has changed too much for this to be rebased
